include/hc.hpp

```diff
@@ -117,6 +451,188 @@ inline uint64_t get_tick_frequency() {
+        : queue_{std::move(queue)} {}
+
+    template<typename InputIterator>
+    completion_future create_blocking_marker_(InputIterator f,
+                                              InputIterator l,
+                                              memory_scope scope,
+                                              std::true_type) const
+    {   // Precondition: readable_bounded_range(f, l) && non_empty(f, l).
+        static_assert(std::is_same<typename std::iterator_traits<InputIterator>::value_type,
+                                   completion_future>::value,
+                      "This requires accelerator_view arguments.");
+
+        std::shared_ptr<Kalmar::KalmarAsyncOp> deps[max_async_ops];
+
+        // since HC sets the barrier bit in each AND barrier packet, we know
+        // the barriers will execute in-order.
+        while (true) {
+            std::ptrdiff_t n = 0;
+            do {
+                // TODO: should not use completion_future implementation detail.
+                deps[n] = f->async_op_;
+                ++n;
+                ++f;
+            } while (f != l && n != max_async_ops);
+
+            if (f == l) {
+                return completion_future{queue_->EnqueueMarkerWithDependency(n,
+                                                                             deps,
+                                                                             scope)};
+            }
+
+            queue_->EnqueueMarkerWithDependency(n, deps, scope);
+        }
+    }
+
+    template<typename InputIterator>
+    completion_future create_blocking_marker_(InputIterator f,
+                                              InputIterator l,
+                                              memory_scope scope,
+                                              std::false_type) const
+    {   // Precondition: readable_bounded_range(f, l) && non_empty(f, l).
+        static_assert(std::is_same<typename std::iterator_traits<InputIterator>::value_type,
+                                   accelerator_view>::value,
+                      "This requires accelerator_view arguments.");
+
+        std::shared_ptr<Kalmar::KalmarAsyncOp> deps[max_async_ops];
+
+        // since HC sets the barrier bit in each AND barrier packet, we know
+        // the barriers will execute in-order.
+        while (true) {
+            std::ptrdiff_t n = 0;
+            do {
+                // TODO: should not use completion_future implementation detail.
+                if (!f->queue_->isEmpty()) {
+                    deps[n] = f->queue_->EnqueueMarker(scope);
+                    ++n;
+                    ++f;
+                }
+            } while (f != l && n != max_async_ops);
+
+            if (f == l) {
+                return completion_future{queue_->EnqueueMarkerWithDependency(n,
+                                                                             deps,
+                                                                             scope)};
```

Comment:
Right, this would fail for e.g. a CPUQueue, however I'm not sure we're ever using a CPU device anywhere at this point? If we are, it might be cleaner to provide a token implementation there, perhaps? Or just disable this code for anything but a HSAQueue. It's worthwhile to note that I do not quite believe that create_blocking_marker works with a CPUQueue anyhow, even in its current implementation, therefore we might be straying into uncharted territory.